### PR TITLE
feat(patch): patch support for flow info and job info and refactor patchbuilders for java sdk

### PR DESCRIPTION
**Pull request by RyanHolstien**
## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

**Comment by shirshanka** [comment]
Fix PR title to reflect new expanded scope: Dataset Properties

**Comment by shirshanka** [comment]
@yoonhyejin could you review this doc for consistency with other similar docs?

File: docs/api/tutorials/adding-dataset-custom-properties.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+## Add Custom Properties With Java SDK
+The following code adds custom properties `cluster_name` and `retention_time` to a dataset named `fct_users_deleted` without affecting existing properties.
+
+```java
+{{ inline /metadata-integration/java/examples/src/main/java/io/datahubproject/examples/DatasetCustomProperties.java }}
+```
+
+## Replace Custom Properties With Java SDK
+The following code replaces the current custom properties to a dataset named `fct_users_deleted` with `cluster_name` and `retention_time`. Note, since this
+is utilizing the upsert API this will do a full replace of the Dataset Properties aspect and will not retain other fields unless specified. To just modify custom properties
```
Comment (shirshanka) [inline docs/api/tutorials/adding-dataset-custom-properties.md:42]: Shouldn't we be implementing the `REPLACE` operation in JsonPatch for this? (https://jsonpatch.com/#replace)

File: metadata-ingestion/src/datahub/specific/custom_properties.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+from datahub.metadata.schema_classes import (
+    KafkaAuditHeaderClass,
+    MetadataChangeProposalClass,
+    SystemMetadataClass,
+)
+
+T = TypeVar("T", bound=MetadataPatchProposal)
+
+
+class CustomPropertiesPatchBuilder(Generic[T], MetadataPatchProposal):
```
Comment (shirshanka) [inline metadata-ingestion/src/datahub/specific/custom_properties.py:13]: @hsheth2 : would like to get your input on this design. 
Salient points:

1. CustomPropertiesPatchBuilder adds patches to the parent builder instead of creating its own set of patches.
2. calls build() on the parent.
3. Offers a parent() method to get access to the parent builder if the caller needs it to chain further methods.

Comment (shirshanka) [reply metadata-ingestion/src/datahub/specific/custom_properties.py:13]: This is different from how we implemented fieldPathPatchBuilder ... so would like to first get consensus on what good looks like, and then we can align both builders to the same pattern.

File: smoke-test/tests/patch/test_dataset_patches.py [hunk 3]
```diff
@@ -387,3 +381,136 @@ def test_field_tags_patch(wait_for_healthchecks):
...
+
+        dataset_properties: Optional[DatasetPropertiesClass] = graph.get_aspect(
+            dataset_urn, DatasetPropertiesClass
+        )
+
+        assert dataset_properties
+        assert dataset_properties.name == orig_dataset_properties.name
+        assert dataset_properties.description == orig_dataset_properties.description
+
+        # Patch custom properties along with name
```
Comment (shirshanka) [inline smoke-test/tests/patch/test_dataset_patches.py:489]: @hsheth2 : look at this for examples of how the embedded builder gets used.

File: metadata-integration/java/examples/src/main/java/io/datahubproject/examples/DatasetCustomPropertiesAddRemove.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+   * @param args
+   * @throws IOException
+   * @throws ExecutionException
+   * @throws InterruptedException
+   */
+    public static void main(String[] args) throws IOException, ExecutionException, InterruptedException {
+      MetadataChangeProposal datasetPropertiesProposal = new DatasetPropertiesPatchBuilder()
+          .urn(UrnUtils.toDatasetUrn("hive", "fct_users_deleted", "PROD"))
+          // the operation type doesn't matter here, what matters are addProperty / removeProperty calls below
+          .op(PatchOperationType.ADD)
```
Comment (shirshanka) [inline metadata-integration/java/examples/src/main/java/io/datahubproject/examples/DatasetCustomPropertiesAddRemove.java:34]: @RyanHolstien : looking at this now, the `op` probably doesn't make sense on the top level builder ever..   we should have `add_x`, `remove_x` or `set_x` methods which imply the `op` on the field called `x`. See the Python sdk example.

File: metadata-integration/java/datahub-client/src/main/java/datahub/client/patch/common/CustomPropertiesPatchBuilder.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+  /**
+   * Remove a property from a custom properties field. If the property doesn't exist, this is a no-op.
+   * @param key
+   * @return
+   */
+  public CustomPropertiesPatchBuilder<T> removeProperty(String key) {
+    return removeProperty(key, null);
+  }
+
+  private CustomPropertiesPatchBuilder<T> removeProperty(String key, String value) {
```
Comment (shirshanka) [inline metadata-integration/java/datahub-client/src/main/java/datahub/client/patch/common/CustomPropertiesPatchBuilder.java:46]: @RyanHolstien : curious why you created removeProperty with key and value both needing to get specified? Isn't `removeProperty(String key)` more natural?

File: metadata-io/src/main/java/com/linkedin/metadata/entity/AspectUtils.java [hunk 6]
```diff
@@ -72,20 +72,35 @@ public static Map<Urn, Aspect> batchGetLatestAspect(
...
-      // Additional changes should never be set as PATCH, if a PATCH is coming across it should be an UPSERT
-      if (ChangeType.PATCH.equals(proposal.getChangeType())) {
-        proposal.setChangeType(ChangeType.UPSERT);
-      }
-      proposal.setAspect(genericAspect);
-      proposal.setAspectName(aspectName);
-      return proposal;
-    } catch (CloneNotSupportedException e) {
-      log.error("Issue while generating additional proposals corresponding to the input proposal", e);
+    MetadataChangeProposal proposal = new MetadataChangeProposal();
```
Comment (shirshanka) [inline metadata-io/src/main/java/com/linkedin/metadata/entity/AspectUtils.java:75]: Why were these changes needed for this PR?

Comment (RyanHolstien) [reply metadata-io/src/main/java/com/linkedin/metadata/entity/AspectUtils.java:75]: It's a long time bug that surfaces due to the changes made here. Basically what happens here is that the underlying datamap gets reused from the change proposal and certain fields get passed through that cause issues. This change removes that bug and allows the patches to go through whenever it's the first modification to the aspect.

File: metadata-integration/java/examples/src/main/java/io/datahubproject/examples/DatasetCustomPropertiesAddRemove.java (hunk 4 above)
Comment (RyanHolstien) [reply metadata-integration/java/examples/src/main/java/io/datahubproject/examples/DatasetCustomPropertiesAddRemove.java:34]: The idea here was to limit what was possible to make the interface simpler. If you want to do a different operation you make a new proposal rather than doing adds, replaces, and deletes within the same one as it's really easy to shoot yourself in the foot and create an invalid patch. It also makes it so you don't need a patch builder style interface per every field and instead it's at the aspect level.

I think it's a bit more intuitive and usable this way without having to read the full JsonPatch docs since this is targeted at people who wouldn't want to construct one themselves since that is also an option on the raw ingestProposal endpoint, but I'm open to rethinking the interface if we think it's too restrictive.

File: docs/api/tutorials/adding-dataset-custom-properties.md (hunk 1 above)
Comment (RyanHolstien) [reply docs/api/tutorials/adding-dataset-custom-properties.md]: Replace is supported, but it's kind of complicated to work with. I prefer pointing people to doing delete -> add

File: metadata-integration/java/datahub-client/src/main/java/datahub/client/patch/common/CustomPropertiesPatchBuilder.java (hunk 5 above)
Comment (RyanHolstien) [reply metadata-integration/java/datahub-client/src/main/java/datahub/client/patch/common/CustomPropertiesPatchBuilder.java:46]: Yeah this doesn't really need to be split out to a separate method, I think when I was writing it I thought this was more readable, but it's kind of less clear. I'm okay with changing it to just be in the public method

**Review by david-leifker — APPROVED** [review]

File: metadata-integration/java/examples/src/main/java/io/datahubproject/examples/DatasetCustomPropertiesAddRemove.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+   * affecting any other properties
+   * @param args
+   * @throws IOException
+   * @throws ExecutionException
+   * @throws InterruptedException
+   */
+    public static void main(String[] args) throws IOException, ExecutionException, InterruptedException {
+      MetadataChangeProposal datasetPropertiesProposal = new DatasetPropertiesPatchBuilder()
+          .urn(UrnUtils.toDatasetUrn("hive", "fct_users_deleted", "PROD"))
+          // the operation type doesn't matter here, what matters are addProperty / removeProperty calls below
```
Comment (shirshanka) [inline metadata-integration/java/examples/src/main/java/io/datahubproject/examples/DatasetCustomPropertiesAddRemove.java:32]: comment is no longer needed

File: metadata-integration/java/datahub-client/src/main/java/datahub/client/patch/datajob/DataJobInfoPatchBuilder.java [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+  }
+
+  @Override
+  public DataJobInfoPatchBuilder removeCustomProperty(@Nonnull String key) {
+    customPropertiesPatchBuilder.removeProperty(key);
+    return this;
+  }
+
+  @Override
+  public DataJobInfoPatchBuilder replaceCustomProperties(Map<String, String> properties) {
```
Comment (shirshanka) [inline metadata-integration/java/datahub-client/src/main/java/datahub/client/patch/datajob/DataJobInfoPatchBuilder.java:129]: change this to `setCustomProperties`

File: metadata-integration/java/datahub-client/src/main/java/datahub/client/patch/dataset/DatasetPropertiesPatchBuilder.java [hunk 9]
```diff
@@ -1,76 +1,119 @@
...
 
-    return value;
+  @Override
+  public DatasetPropertiesPatchBuilder removeCustomProperty(@Nonnull String key) {
+    this.customPropertiesPatchBuilder.removeProperty(key);
+    return this;
+  }
+
+  @Override
+  public DatasetPropertiesPatchBuilder replaceCustomProperties(Map<String, String> properties) {
```
Comment (shirshanka) [inline metadata-integration/java/datahub-client/src/main/java/datahub/client/patch/dataset/DatasetPropertiesPatchBuilder.java:99]: change to `setCustomProperties`

Comment (shirshanka) [reply metadata-integration/java/datahub-client/src/main/java/datahub/client/patch/dataset/DatasetPropertiesPatchBuilder.java:99]: write javadoc for what this does and when to use it versus addCustomProperty / removeCustomProperty

**Review by shirshanka — APPROVED** [review]
LGTM!
